django/db/models/aggregates.py

```diff
@@ -13,12 +14,35 @@
 class Aggregate(Func):
     contains_aggregate = True
     name = None
+    filter_template = '%s FILTER (WHERE %%(filter)s)'
+
+    def __init__(self, *args, filter=None, **kwargs):
+        self.filter = filter
+        super().__init__(*args, **kwargs)
+
+    def get_source_fields(self):
+        # We're not interested in the filter expression.
+        return [e._output_field_or_none for e in super().get_source_expressions()]
+
+    def get_source_expressions(self):
+        source_expressions = super(Aggregate, self).get_source_expressions()
+        if self.filter:
+            source_expressions += [self.filter]
+        return source_expressions
+
+    def set_source_expressions(self, exprs):
+        if self.filter:
+            self.filter = exprs[-1]
+            exprs = exprs[:-1]
+        return super(Aggregate, self).set_source_expressions(exprs)
 
     def resolve_expression(self, query=None, allow_joins=True, reuse=None, summarize=False, for_save=False):
         # Aggregates are not allowed in UPDATE queries, so ignore for_save
         c = super().resolve_expression(query, allow_joins, reuse, summarize)
+        if c.filter:
+            c.filter = c.filter.resolve_expression(query, allow_joins, reuse, summarize)
         if not summarize:
-            expressions = c.get_source_expressions()
+            expressions = super(Aggregate, c).get_source_expressions()
```

Comment:
We might want to mention we use `super()` to avoid checking against `self.filter`.